examples/fluids/src/dm_utils.c

```diff
@@ -0,0 +1,352 @@
+                                  CeedInt num_fields, const CeedInt *field_sizes, DM dm) {
+  PetscInt  dim, q_order = degree + q_extra;
+  PetscBool is_simplex = PETSC_TRUE;
+  PetscFE   fe;
+  MPI_Comm  comm = PetscObjectComm((PetscObject)dm);
+
+  PetscFunctionBeginUser;
+  PetscCall(DMPlexIsSimplex(dm, &is_simplex));
+
+  // Setup DM
+  PetscCall(DMGetDimension(dm, &dim));
+  for (PetscInt i = 0; i < num_fields; i++) {
+    PetscFE  fe_face;
+    PetscInt q_order = degree + q_extra;
+
+    PetscCall(PetscFECreateLagrange(comm, dim, field_sizes[i], is_simplex, degree, q_order, &fe));
+    if (setup_faces) PetscCall(PetscFEGetHeightSubspace(fe, 1, &fe_face));
+    PetscCall(DMAddField(dm, NULL, (PetscObject)fe));
+    PetscCall(PetscFEDestroy(&fe));
+  }
+  PetscCall(DMCreateDS(dm));
+
+  // Project coordinates to enrich quadrature space
+  if (setup_coords) {
+    DM             dm_coord;
+    PetscDS        ds_coord;
+    PetscFE        fe_coord_current, fe_coord_new, fe_coord_face_new;
+    PetscDualSpace fe_coord_dual_space;
+    PetscInt       fe_coord_order, num_comp_coord;
+
+    PetscCall(DMGetCoordinateDM(dm, &dm_coord));
+    PetscCall(DMGetCoordinateDim(dm, &num_comp_coord));
+    PetscCall(DMGetRegionDS(dm_coord, NULL, NULL, &ds_coord, NULL));
+    PetscCall(PetscDSGetDiscretization(ds_coord, 0, (PetscObject *)&fe_coord_current));
+    PetscCall(PetscFEGetDualSpace(fe_coord_current, &fe_coord_dual_space));
+    PetscCall(PetscDualSpaceGetOrder(fe_coord_dual_space, &fe_coord_order));
+
+    // Create FE for coordinates
+    PetscCheck(fe_coord_order == 1 || coord_order == 1, PetscObjectComm((PetscObject)dm), PETSC_ERR_USER_INPUT,
+               "Only linear mesh geometry supported. Recieved %d order", fe_coord_order);
+    PetscCall(PetscFECreateLagrange(comm, dim, num_comp_coord, is_simplex, fe_coord_order, q_order, &fe_coord_new));
+    if (setup_faces) PetscCall(PetscFEGetHeightSubspace(fe_coord_new, 1, &fe_coord_face_new));
+    PetscCall(DMProjectCoordinates(dm, fe_coord_new));
+    PetscCall(PetscFEDestroy(&fe_coord_new));
+  }
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/**
+  @brief Finish setting up `DM`
+
+  To be called after `DMSetupByOrder_FEM` and all strong BCs have be declared via `DMAddBoundaries`.
+
+  @param[in]   setup_coords    Flag to setup coordinate spaces
+  @param[out]  dm              `DM` to setup
+
+  @return An error code: 0 - success, otherwise - failure
+**/
+PetscErrorCode DMSetupByOrder_Finalize(PetscBool setup_coords, DM dm) {
+  PetscBool is_simplex;
+  PetscFunctionBeginUser;
+
+  PetscCall(DMPlexIsSimplex(dm, &is_simplex));
+  // Set tensor permutation if needed
+  if (!is_simplex) {
+    PetscCall(DMPlexSetClosurePermutationTensor(dm, PETSC_DETERMINE, NULL));
+    if (setup_coords) {
+      DM dm_coord;
+
+      PetscCall(DMGetCoordinateDM(dm, &dm_coord));
+      PetscCall(DMPlexSetClosurePermutationTensor(dm_coord, PETSC_DETERMINE, NULL));
+    }
+  }
```

Comment:
I worry that this is really easy to forget and is a potential hard-to-find-bug bomb. But I'm not seeing an obvious workaround if you aren't setting any Dirichlet BCs inside `DMSetupByOrder_FEM` like we do in Ratel.

Reply:
Yeah, I will want to move to calling `DMAddBoundary` within `DMSetupByOrder_FEM`, but once I go through with the boundary condition refactor. 

The main issue is that the `DMAddBoundary` calls requires other structs that might be be available everywhere `DMSetupByOrder_FEM` needs to be called. Otherwise I'd just add them in and call it good enough for now.

Reply:
Addressed in this thread: https://github.com/CEED/libCEED/pull/1308#discussion_r1306219024